aopy/data/bmi3d.py

```diff
@@ -197,6 +197,149 @@ def proc_ecube_data(data_path, data_source, result_filepath, result_name='broadb
+        result_filepath (str): Path to save the filtered LFP data.
+        mean_subtract (bool, optional): Whether to subtract the mean from the filtered LFP signal.
+                                        Default is True.
+        dtype (str, optional): Data type for the filtered LFP signal. Default is 'int16'.
+        max_memory_gb (float, optional): Maximum memory (in gigabytes) to use for filtering. Default is 1.0 GB.
+        **filter_kwargs: Additional keyword arguments to customize the filtering process.
+                        These arguments will be passed to the filtering function.
+
+    Raises:
+        IOError: If the input eCube recording file is not found.
+        MemoryError: If the specified max_memory_gb is insufficient for the filtering process.
+
+    Note:
+        This function is used in the :func:`~aopy.preproc.warppers.proc_lfp` wrapper.
+    '''
+    lfp_samplerate = filter_kwargs.pop('lfp_samplerate', 1000)
+
+    metadata = load_ecube_metadata(ecube_filepath, 'Headstages')
+    samplerate = metadata['samplerate']
+    downsample_factor = int(samplerate/lfp_samplerate)
+    n_channels = int(metadata['n_channels'])
+    n_samples = int(metadata['n_samples'])
+
+    # Figure out how much data we can load at once
+    max_samples = int(max_memory_gb * 1e9 / np.dtype(dtype).itemsize)
+    chunksize = int(max_samples / metadata['n_channels'])
+    n_whole_chunks = int(n_samples / chunksize)
+    n_remaining = n_samples - (chunksize * n_whole_chunks)
+    lfp_samples = np.ceil(chunksize/downsample_factor) * n_whole_chunks + np.ceil(n_remaining/downsample_factor)
+    print("lfp chunks should be ", np.ceil(chunksize/downsample_factor))
+    print("last chunk should be ", np.ceil(n_remaining/downsample_factor))
+    print("total samples: ", lfp_samples)
+
+    # Create an hdf dataset
+    hdf = h5py.File(result_filepath, 'a') # should append existing or write new?
+    dset = hdf.create_dataset('lfp_data', (lfp_samples, n_channels), dtype=dtype)
+
+    # Filter broadband data into LFP directly into the hdf file
+    n_lfp_samples = 0
+    for broadband_chunk in load_ecube_data_chunked(ecube_filepath, 'Headstages', chunksize=chunksize):
+        lfp_chunk = precondition.filter_lfp(broadband_chunk, samplerate, **filter_kwargs)
+        chunk_len = lfp_chunk.shape[0]
+        print(n_lfp_samples, n_lfp_samples+chunk_len, lfp_samples)
+        dset[n_lfp_samples:n_lfp_samples+chunk_len,:] = lfp_chunk
+        n_lfp_samples += chunk_len
+        
+    if mean_subtract:
+        dset -= np.mean(dset, axis=0, dtype=dtype) # hopefully this isn't constrained by memory
+    hdf.close()
+
+    # Append the lfp metadata to the file
+    lfp_metadata = metadata
+    lfp_metadata['lfp_samplerate'] = lfp_samplerate # for backwards compatibility
+    lfp_metadata['samplerate'] = lfp_samplerate
+    lfp_metadata['n_samples'] = lfp_samples
+    lfp_metadata['low_cut'] = 500
+    lfp_metadata['buttord'] = 4
```

Comment:
Will these always be 500 and 4?

Reply:
no, those are the defaults, but the line below should update them with user input